Move xray propagator from contrib (no history)

martinkuba opened the pull request:
This is a second alternative to https://github.com/open-telemetry/opentelemetry-js/pull/4544. It adds the [opentelemetry-propagator-aws-xray](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/propagators/opentelemetry-propagator-aws-xray) package from contrib without bringing over any history.

---
part of https://github.com/open-telemetry/opentelemetry-js/issues/4494

It should be possible to configure the `xray` propagator using the `OTEL_PROPAGATORS` env variable. This is currently documented [here](https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_propagators), and it is also part of the [lambda spec](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/faas/aws-lambda.md#xray-lambda-propagator-configuration).

pichlermarc commented on packages/propagator-aws-xray/package.json:
```diff
@@ -0,0 +1,83 @@
+    "@types/mocha": "10.0.6",
+    "@types/node": "18.6.5",
+    "@types/webpack-env": "1.16.3",
+    "babel-plugin-istanbul": "6.1.1",
```
used in scripts
```suggestion
    "babel-plugin-istanbul": "6.1.1",
    "codecov": "3.8.3",
```

martinkuba replied:
added

pichlermarc commented on packages/propagator-aws-xray/package.json:
```diff
@@ -0,0 +1,83 @@
+    "karma-coverage": "2.2.1",
+    "karma-mocha": "2.0.1",
+    "karma-spec-reporter": "0.0.36",
+    "karma-webpack": "4.0.2",
```
used in scripts.
```suggestion
    "karma-webpack": "4.0.2",
    "lerna": "6.6.2",
```

martinkuba replied:
added

pichlermarc reviewed (approved):
Looks good :+1: We may need to add/remove some `devDependencies` (edit: for the changelog, we could mention that it has moved over, and that versions 1.4 through 1.22 have been skipped)

Note to reviewers: this package is already stable in contrib, so we can only make minimal changes to it.

martinkuba commented on packages/propagator-aws-xray/package.json:
```diff
@@ -0,0 +1,83 @@
+    "karma-webpack": "4.0.2",
+    "mocha": "10.2.0",
+    "nyc": "15.1.0",
+    "rimraf": "5.0.5",
```
removed

martinkuba commented:
> ... for the changelog, we could mention that it has moved over, and that versions 1.4 through 1.22 have been skipped

I have added this as a feature to changelog. Let me know if you want to classify it differently. And I have added a note in the REAME of the package about the skipped versions.

weyert commented:
Does that mean we could move Google Trace to here too? Would be nice to be able to configure that too

pichlermarc commented:
@martinkuba tests are likely failing due to a different version of `@grpc/gprc-js` being pulled in.
I'll push some changes to package-lock.json to this PR in a moment that should resolve the build issues.

We'll need to update `@grpc/grpc-js` and the tests for the exporters in another PR.

trentm reviewed (approved):
LGTM, with a nit in the rEADME.

anuraags commented:
Is there any movement happening on this? Would like to be able to use an official xray propagator in my services.

pichlermarc commented:
> Is there any movement happening on this? Would like to be able to use an official xray propagator in my services.

It will be included in the next release `1.24.0`.

Flarna commented:
> Would like to be able to use an official xray propagator in my services

I don't think it's more official than now just because the repo changes.

martinkuba commented:
@anuraags You can use the xray propagator now - the [lambda instrumentation](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-aws-lambda/README.md) uses it automatically as well as the [lambda layer](https://github.com/open-telemetry/opentelemetry-lambda/tree/main/nodejs). You can also [configure it manually](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-sdk-trace-base/src/types.ts#L60) in the TracerProvider. This change will only make it possible to set the propagator via the env variable.

weyert commented:
> > Does that mean we could move Google Trace to here too? Would be nice to be able to configure that too
> 
> No. It's not a well-known propagator according to [the specification.](https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_propagators)

Curious why only one of the major Cloud providers gets special treatment within the OpenTelemetry specification

Flarna commented:
I guess it's up to other cloud providers (or their users) to create PRs in OTel spec repo to get their special thingies in.

martinkuba commented:
@weyert It's based on the list of known values for the OTEL_PROPAGATORS env variable defined [here](https://opentelemetry.io/docs/languages/sdk-configuration/general/#otel_propagators). But I don't think there is anything preventing us from adding others.

And as a side note, the intent of this effort is to actually move the xray propagation from lambda instrumentations to the global configuration (via the env var), as having it hard-coded gave x-ray special treatment.
